Handle IPSL-CM6 (#1153)
ESMValGroup/ESMValCore: doc/develop/fixing_data.rst

.. _fixing_data:

***********
Fixing data
***********

The baseline case for ESMValCore input data is CMOR fully compliant

> data that is read using
> `Iris load_raw function <https://scitools-iris.readthedocs.io/en/latest/generated/api/iris.html#iris.load_raw>`_.

Reviewer: Thanks to intersphinx, we can write this much more compact as
```suggestion
data that is read using Iris' :func:`iris:iris.load_raw`.
```
For more information see [intersphinx](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) and [roles](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cross-referencing-python-objects).

> ESMValCore also allows for some departures with compliance (see

Reviewer: departure *from* compliance

:ref:`cmor_check_strictness`). Beyond that situation, some datasets
(either model or observations) contain (known) errors that would
normally prevent them from being processed. The issues can be in
the metadata describing the dataset and/or in the actual data.
Typical examples of such errors are missing or wrong attributes (e.g.
attribute ''units'' says 1e-9 but data are actually in 1e-6), missing or
mislabeled coordinates (e.g. ''lev'' instead of ''plev'' or missing
coordinate bounds like ''lat_bnds'') or problems with the actual data
(e.g. cloud liquid water only instead of sum of liquid + ice as specified by the CMIP data request).

> As an extreme case, some others data sources simply are not NetCDF

Reviewer: ```suggestion
 As an extreme case, some data sources simply are not NetCDF
```

> files and must go through other data load function.

Reviewer: ```suggestion
files and must go through some other data load function.
```

The ESMValCore can apply on the fly fixes to such datasets when
issues can be fixed automatically. This is implemented for a set

> of `Natively supported non-CMIP datasets`_. The following provide

Reviewer: ```suggestion
of `Natively supported non-CMIP datasets`_. The following provides
```

.. _fixing_data:

******************************************

> Adapting to data sources

Reviewer 2: I agree that the title could be improved, but I would prefer to keep the words 'fix' and `dataset' in the title, because those are the words that everyone uses when talking about this topic.

Author: OK

******************************************

> The baseline case for ESMValTool input data is CMOR fully compliant

Reviewer 2: ```suggestion
The baseline case for ESMValCore input data is CMOR fully compliant
```
please use the word ESMValCore everywhere in the ESMValCore documentation.

Author: Done.

> data that is read using Iris load fuction. ESMValTool also allows for

Reviewer 2: ```suggestion
data that is read using Iris load function. ESMValCore also allows for
```
and could you make the iris load function a link to the iris documentation for added clarity?

Author: Done

Natively supported non-CMIP datasets
====================================

Some fixed datasets and native models formats are supported through
the ``native6`` project or through a dedicated project.

.. _fixing_native_models:

Native models : IPSL-CM6,... 
-----------------------------

The following models are natively supported through the procedure
described above (:ref:`fix_structure`) and at
:ref:`configure_native_models`:

  - **IPSL-CM6** : both output formats (i.e. the ``Output`` and the
    ``Analyse / Time series`` formats) are supported, and should be
    configured in recipes as e.g.:

    .. code-block:: yaml

      datasets:
        - {simulation: CM61-LR-hist-03.1950, exp: piControl, freq: Analyse/TS_MO,
           account: p86caub,  status: PROD, dataset: IPSL-CM6, project: native6,

> 	   root: /thredds/tgcc/store} 

Reviewer 2: Is `root` describing some input path? Is that still needed with the addition of the extra facets file?

Author: Yes, that is still needed because , for model development use case, modelers have some choice on where to put the data, which may change from one experiment to the other

        - {simulation: CM61-LR-hist-03.1950, exp: historical, freq: Output/MO,
           account: p86caub,  status: PROD, dataset: IPSL-CM6, project: native6,
	   root: /thredds/tgcc/store} 

    The ``Output`` format is an example of a case where variables are
    grouped in multi-variable files, which name cannot be computed
    directly from datasets attributes alone but requires an

>     ref:`Extra_facets file <extra_facets>`. These multi-variable

Reviewer 2: Link is broken

Author: Because #1124 is not yet merged. I removed the link. To be polished for V2.4

Author: Yet done

    files must also undergo some data selection, which may involve an
    external process for performance purpose.

> ERA5 and MSWEP datasets

Reviewer 2: Is the text on ERA5 and MSEP changed? If not, would it be possible to keep the original formatting so GitHub understands that it is the same?

Author: I tried

Author: Because that text is both moved and differently indented , GitHub doesn't understand thta it is the same

Natively supported non-CMIP datasets
====================================

> Some fixed datasets and native models formats are supported through

Reviewer: ```suggestion
Some fixed datasets and native model formats are supported through
```

Reviewer: Do you have reservations about this change?

For more info: http://www.gloh2o.org/

.. _fixing_native_models:

Native models : IPSL-CM6,...
-----------------------------

The following models are natively supported through the procedure
described above (:ref:`fix_structure`) and at
:ref:`configure_native_models`:

>   - **IPSL-CM6** : both output formats (i.e. the ``Output`` and the

Reviewer: I suggest to move to a sub-section style approach also for the models. In addition to the following suggestion, this requires a change of indentation of the whole IPSL-CM6 part.

```suggestion
IPSL-CM6
~~~~~~~~
Both output formats (i.e. the ``Output`` and the
```

    ``Analyse / Time series`` formats) are supported, and should be
    configured in recipes as e.g.:

    .. code-block:: yaml

      datasets:
        - {simulation: CM61-LR-hist-03.1950, exp: piControl, freq: Analyse/TS_MO,
           account: p86caub,  status: PROD, dataset: IPSL-CM6, project: IPSLCM,
           root: /thredds/tgcc/store}
        - {simulation: CM61-LR-hist-03.1950, exp: historical, freq: Output/MO,
           account: p86caub,  status: PROD, dataset: IPSL-CM6, project: IPSLCM,
           root: /thredds/tgcc/store}

>     .. _ipslcm_extra_facets_example:

Reviewer: Is this intended as a label for the yaml block? In that case I think it has to go before the block, i.e.
```
.. _ipslcm_extra_facets_example:

.. code-block:: yaml
```

Author: No, it is a ref for the text below

    The ``Output`` format is an example of a case where variables are
    grouped in multi-variable files, which name cannot be computed
    directly from datasets attributes alone but requires to use an
    extra_facets file, which principles are explained in
    :ref:`extra_facets`, and which content is :download:`available here
    </../esmvalcore/_config/extra_facets/ipslcm-mappings.yml>`. These

>     multi-variable files must also undergo some data selection, which
>     may involve an external process for performance purpose.

Reviewer: I don't think we need to mention CDO here.
```suggestion
    multi-variable files must also undergo some data selection.
```
